Code/Tools/AssetProcessor/native/ui/ProductDependencyTreeModel.cpp

```diff
@@ -0,0 +1,358 @@
+        }
+
+        if (!parentItem)
+        {
+            return QModelIndex();
+        }
+
+        ProductDependencyTreeItem* childItem = parentItem->GetChild(row);
+
+        if (childItem)
+        {
+            QModelIndex index = createIndex(row, column, childItem);
+            Q_ASSERT(checkIndex(index));
+            return index;
+        }
+        return QModelIndex();
+    }
+
+    int ProductDependencyTreeModel::rowCount(const QModelIndex& parent) const
+    {
+        if (parent.column() > 0)
+        {
+            return 0;
+        }
+
+        ProductDependencyTreeItem* parentItem = nullptr;
+        if (!parent.isValid())
+        {
+            parentItem = m_root.get();
+        }
+        else
+        {
+            parentItem = static_cast<ProductDependencyTreeItem*>(parent.internalPointer());
+        }
+
+        if (!parentItem)
+        {
+            return 0;
+        }
+        return parentItem->getChildCount();
+    }
+
+    int ProductDependencyTreeModel::columnCount(const QModelIndex& /*parent*/) const
+    {
+        return static_cast<int>(ProductDependencyTreeColumns::Max);
+    }
+
+    QVariant ProductDependencyTreeModel::data(const QModelIndex& index, int role) const
+    {
+        if (!index.isValid())
+        {
+            return QVariant();
+        }
+
+        ProductDependencyTreeItem* item = static_cast<ProductDependencyTreeItem*>(index.internalPointer());
+        switch (role)
+        {
+        case Qt::DisplayRole:
+            return item->GetDataForColumn(index.column());
+        case Qt::DecorationRole:
+            // Only show the icon in the name column
+            if (index.column() == static_cast<int>(AssetTreeColumns::Name))
+            {
+                return m_fileIcon;
+            }
+            break;
+        case Qt::ToolTipRole:
+            // Purposely return an empty string, so mousing over rows clear out.
+            return QString("");
+        }
+        return QVariant();
+    }
+
+    bool ProductDependencyTreeModel::setData(const QModelIndex& /*index*/, const QVariant& /*value*/, int /*role*/)
+    {
+        return false;
+    }
+
+    Qt::ItemFlags ProductDependencyTreeModel::flags(const QModelIndex& index) const
+    {
+        return Qt::ItemIsSelectable | QAbstractItemModel::flags(index);
+    }
+
+    QModelIndex ProductDependencyTreeModel::parent(const QModelIndex& index) const
+    {
+        if (!index.isValid())
+        {
+            return QModelIndex();
+        }
+
+        ProductDependencyTreeItem* childItem = static_cast<ProductDependencyTreeItem*>(index.internalPointer());
+        ProductDependencyTreeItem* parentItem = childItem->GetParent();
+
+        if (parentItem == m_root.get() || parentItem == nullptr)
+        {
+            return QModelIndex();
+        }
+        QModelIndex parentIndex = createIndex(parentItem->GetRow(), 0, parentItem);
+        Q_ASSERT(checkIndex(parentIndex));
```

Comment:
I would like to caution on the use of Assert() inside of the Asset Processor. If this happens for whatever reason then the AP will suddenly die without any explanation to the user. If the user is using the Editor, then the AP will just "stop working"

Reply:
I see. What would you suggest as a better approach? They are all checking index generated by createIndex(). I am thinking of returning an (invalid) index by `return QModelIndex()`.

Reply:
finished